summary.Rmd

```diff
@@ -2,12 +2,40 @@ Summary
 =======
 
 1. Use available tools
+    a. Image registries (e.g. Docker Hub),
+    b. Dockerfile generators (e.g. repo2docker)
+    c. Project templates (e.g. cookie cutter) could go here and one Dockerfile per project?
+    d. Linters
 2. Use versioned images
+    a. Mention specific sha digests?
 3. Format for clarity
+    a. Use indentation, line breaks and unabbreviated arguments
+    b. Clarity > brevity
+    c. Group related commands
 4. Document within the Dockerfile
-5. Order the instructions
+    a. Add comments explaining the steps
+    b. Use labels for metadata
+    c. Include build and run commands
 6. Specify software versions
+    a. Pin system library versions
+    b. Explictly list package versions to install
+    c. Comment why each is needed
+9. Use version control
+    a. Include CI testing
+    b. Use one Dockerfile per project
```

Comment:
Well I think it could go under either rule (but probably only one) so have left it here and put it under Rule 1 for discussion.

Reply:
I see what you mean - the use of templates is definitely better fitting under using available tools, but I'm not sure about the point to use one Dockerfile for project, it doesn't fit nicely with either, and I can think of many projects I've done that have more than one Dockerfile. It's logical to have one Dockerfile for a small, well scoped project, but if you are releasing containers for a tool for different use cases, it's hard to create just one to fit the bill. Also if there is any kind of orchestration, the one Dockerfile rule just won't work. Do we perhaps want to remove it from this master list, and make a much more general statement, e.g.,

> In the case of having a smaller, well scoped project, it might be appropriate to try and develop one main Dockerfile to reproduce the work. 

The example I'm trying to think of (but am having trouble) are cases when one Dockerfile was appropriate but the researcher made more than one.

Reply:
Agreed - I think that would be better. In general, I'm in favour of reducing the text to help distill out the most important rules, as currently I feel some of them are getting diluted with a little too much discussion of technicalities and edge cases. The more general statement would help pare down the main text and we could certainly remove it from the summary list.